### ISSUE TITLE
Help needed

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/pixel-level-contrastive-learning#1]
Need someone to review https://github.com/lucidrains/pixel-level-contrastive-learning/blob/main/pixel_level_contrastive_learning/pixel_level_contrastive_learning.py#L305

The distance threshold in the paper was said to be `0.7`, but that did not work for me (resulted in nearly all positive pixel pairs).

![Screenshot from 2020-11-24 16-25-56](https://user-images.githubusercontent.com/108653/100166513-c074e300-2e71-11eb-85e9-a64e94da2fe1.png)

**Comment by lucidrains** [comment]
![Screenshot from 2020-11-24 16-31-52](https://user-images.githubusercontent.com/108653/100166894-a556a300-2e72-11eb-9e1d-c0a00ecf5720.png)

**Comment by conradry** [comment]
To me, the size of a "feature map bin" is equal the downsampling factor of the model; i.e. for a resnet each bin is 32x32. The diagonal length of the bin would be ~45 and the threshold distance (0.7 * 45) would be  ~32. It at least sounds plausible, but the paper isn't very clear.

**Comment by georgel2004** [comment]
See if this helps - Dense Contrastive Learning for Self-Supervised Visual Pre-Training - https://arxiv.org/pdf/2011.09157.pdf

**Comment by WilhelmT** [comment]
Hi,

In the ablation study, the threshold is as high as 2.8, which according to your interpretation would cover the whole image multiple times over. 

This is my interpretation;
T=0.7 means that two pixels in feature maps of different scales, both warped to the feature map of the original image, have to be within a distance of 0.7. This is how distance in the original feature map space is defined; by normalization I think the authors simply mean that a feature map pixel for the original image is defined to a have diagonal length of 1 (1/sqrt(2) in width and height).

**Comment by conradry** [comment]
@WilhelmT I'm not sure that I understand your comment, so maybe this is redundant, but I think the normalization distances are equivalent to the distances I've marked with red and blue arrows below:

<img width="492" alt="bin_width" src="https://user-images.githubusercontent.com/20257674/102030316-540a4700-3d80-11eb-86bb-d862aab9c81f.png">

This means that you calculate pairwise distances between views once but normalize the distances differently if you're comparing view1 to view2 or comparing view2 to view1.

**Comment by lucidrains** [comment]
ok, I decided to go with the assumption of having the diagonal of each 'pixel' in the feature map normalized to 1

https://github.com/lucidrains/pixel-level-contrastive-learning/commit/8fa3425a243eb4892cfa5a7b453c616370fae2c2
